ocular/src/keyring.rs

```diff
@@ -0,0 +1,752 @@
+
+        // Check if new key exists
+        if self.key_exists(new_name)? && !override_if_exists {
+            eprintln!("New key name '{}', already exists.", new_name);
+            return Err(KeyStoreError::Exists(new_name.to_string()));
+        }
+
+        // Proceed with rename
+        let current_name = &KeyName::new(current_name)
+            .unwrap_or_else(|_| panic!("Could not create KeyName for '{}'.", current_name));
+        let new_name = &KeyName::new(new_name)
+            .unwrap_or_else(|_| panic!("Could not create KeyName for '{}'.", new_name));
+
+        let key = self
+            .key_store
+            .as_ref()
+            .expect("Error accessing key store.")
+            .load(current_name)
+            .expect("Could not load key.");
+
+        // Create new key.
+        let _result = self
+            .key_store
+            .as_ref()
+            .expect("Error accessing key store.")
+            .store(new_name, &key);
+
+        // Delete old key.
+        let _result = self.delete_key(current_name);
+
+        Ok(())
+    }
+
+    fn get_key_address(&self, name: &str) -> Result<PublicKeyOutput, KeyStoreError> {
+        // Check if key exists
+        if !self.key_exists(name)? {
+            eprintln!("Key '{}', does not exist.", name);
+            return Err(KeyStoreError::DoesNotExist(name.to_string()));
+        }
+
+        // Prepare key and parse into pem
+        let key_name = &KeyName::new(name)
+            .unwrap_or_else(|_| panic!("Could not create KeyName for '{}'.", name));
+        let key = self
+            .key_store
+            .as_ref()
+            .expect("Error accessing key store.")
+            .load(key_name)
+            .expect("Could not load key");
+        let key: k256::elliptic_curve::SecretKey<k256::Secp256k1> = key
+            .to_pem(LineEnding::default())
+            .expect("Could not convert to pem.")
+            .parse()
+            .expect("Could not parse pem");
+
+        // Convert to deepspace private key for address conversion
+        let key_bytes = key.to_bytes();
+
+        // Obtain public key bytes
+        let secp256k1_engine = secp256k1::Secp256k1::new();
+        let secp256k1_secret_key =
+            secp256k1::SecretKey::from_slice(&key_bytes).expect("Could not create secret key.");
+        let secp256k1_public_key =
+            secp256k1::PublicKey::from_secret_key(&secp256k1_engine, &secp256k1_secret_key);
+        let secp256k1_public_key_bytes = secp256k1_public_key.serialize();
+
+        // Obtain address hash
+        let sha256 = Sha256::digest(secp256k1_public_key_bytes);
+        let ripemd160 = Ripemd::digest(&sha256);
+        let mut bytes: [u8; 20] = Default::default();
+        bytes.copy_from_slice(&ripemd160[..]);
+
+        // Finally encode address to bech32
+        // TODO: Support other prefixes
+        let bech32_encoded_address =
+            Bech32::new(COSMOS_ADDRESS_PREFIX.to_string(), bytes.to_base32())
```

Comment:
UX design decision (cont.): thoughts on having prefixes specified at the keychain/config level vs allowing the user to specify per key (i.e. being able to have multiple different prefixes in an individual key store)?

Reply:
One thing I was working on is an `enum` representation of each chain name for convenience to be used with client creation. It will still be optional to pass a `&str` name but an enum gives some extra confidence that there are no typos and that the chain is supported. Maybe something similar would be useful for prefixes?

Reply:
The potential pitfall of such an enum would be that as new chains are created the enum would have to be updated. With the optional string param that would allow for users to pass in their own custom chain names, but I'm not sure how helpful the enum would be if the user can override it regardless at that point. I think it would be best to leave full autonomy to the user so as not to pigeonhole them.

Reply:
That's true. We can stick with string slices for now and if the need arises think about something like this later.

Reply:
> UX design decision (cont.): thoughts on having prefixes specified at the keychain/config level vs allowing the user to specify per key (i.e. being able to have multiple different prefixes in an individual key store)?

On second read I think maybe we're viewing this wrong. The prefix is only for the public representation of the key, it has no effect on how the private key is generated afaik. Therefore, whatever chain the client is configured for should be the prefix that gets used when presenting the public address. i.e. if you created a client for `cosmoshub`, the client's config object should contain a field `prefix` with value `"cosmos"`, no?